File: src/Runtime/Sentence/SentenceExpressionValueArray.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

using System.Collections.Generic;

namespace peak.interpreter {

	class SentenceExpressionValueArray : SentenceExpression {
		private List<SentenceExpression> _expressionArray = null;
		public SentenceExpressionValueArray(List<SentenceExpression> valueExpressionArray) {
			_expressionArray = valueExpressionArray;
		}
		public override ExecuteResult Execute(Space space) {
			var valueArray = new ValueArray();
			if (_expressionArray.Count != 0) {
				valueArray.Reserve(_expressionArray.Count);
				for (var i = 0; i < _expressionArray.Count; ++i) {
					var expression = _expressionArray[i];
					if (!IsSuccess(expression.Execute(space))) {
						ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Array, "The array execute failed!");
						return ExecuteResult.Failed;
					}
					var value = expression.value;
					valueArray.EmplaceBack(value);
				}
			}
			SetValue(valueArray);
			return ExecuteResult.Successed;
		}
	}
} // namespace peak.interpreter

File: src/runtime/sentence/sentence_expression_value_array.cpp
#include "sentence_expression_value_array.h"
#include "runtime/value/value_array.h"

using namespace peak;

SentenceExpressionValueArray::SentenceExpressionValueArray(const std::vector<std::shared_ptr<SentenceExpression>>& valueExpressionArray)
	: _expressionArray(valueExpressionArray) {
}
ExecuteResult SentenceExpressionValueArray::Execute(std::shared_ptr<Space> space) {
	auto valueArray = std::make_shared<ValueArray>();
	if (!_expressionArray.empty()) {
		valueArray->Reserve(_expressionArray.size());
		for (auto i = 0u; i < _expressionArray.size(); ++i) {
			auto expression = _expressionArray[i];
			if (!IsSuccess(expression->Execute(space))) {
				ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Array, "The array execute failed!");
				return ExecuteResult::Failed;
			}
			auto value = expression->GetValue();
			valueArray->EmplaceBack(value);
		}
	}
	SetValue(valueArray);
	return ExecuteResult::Successed;
}